dot/state/roottottrie.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+)
+
+type rootToTrieMap struct {
+	rootToTrie map[common.Hash]*trie.Trie
+	sync.RWMutex
+}
```

Comment:
I'm curious as to the advantage of this over a *sync.Map

Reply:
Just type safety / cleaner code. That's how I found out about the empty pruned keys so I thought I would push it 😄

I also don't think it matters much since we just read and access it not in the million per second AFAIK.
If we identify it as a bottleneck later, then we can change it to a sync.Map, which is really for corner performance cases I would say. Let's not pre-pre-optimize 😉

Reply:
It doesn't suck at garbage collection

![](https://media0.giphy.com/media/acttIrNAHaoco/giphy.gif?cid=ecf05e47ls5yx4tbl9w32lls26wxej1vnkumjxc83whlzurx&rid=giphy.gif&ct=g)